Add 3rd memory region: a global Python stack for scoped allocation

dpgeorge opened the pull request:
This patch introduces the MICROPY_ENABLE_PYSTACK option (disabled by default) which enables a "Python stack" that allows to allocate and free memory in a scoped way, similar to alloca().  When MICROPY_ENABLE_PYSTACK is enabled then alloca() is no longer required or used.

The benefits of enabling this option are:
- Toolchains without alloca() can use this feature to obtain correct and efficient scoped memory allocation (compared to using the heap instead of alloca(), which is slower).  The heapless properties of uPy still hold without the use of alloca (eg you can call functions without allocating on the heap).
- Even if alloca() is available, enabling the Py-stack gives slightly more efficient use of stack space when calling nested Python functions, due to the way that compilers implement alloca().
- Enabling the Py-stack with the stackless mode allows for even more efficient stack usage, as well as retaining high performance (because the heap is no longer used to build and destroy stackless code states).
- With Py-stack and stackless enabled, Python-calling-Python is no longer recursive in the C mp_execute_bytecode function.
- With Py-stack enabled it's now possible to do calls of the form `f(*arg)` and `f(**kw)` without allocating on the heap.

The micropython.pystack_use() function is included to measure usage of the Python stack.

To use the Python stack:
- enable MICROPY_ENABLE_PYSTACK compile-time option
- allocate some RAM (statically) for the Python stack region (1-2k bytes seems good)
- call `mp_pystack_init(region_start, region_end)` before calling `mp_init()`
- optionally enable MICROPY_STACKLESS and MICROPY_STACKLESS_STRICT

It may be possible in the future for bare-metal systems to use the other end of the C stack for the Python stack (C stack grows down, Python stack grows up).  This way it wouldn't require to have a separate region for the Python stack and allows to make maximum use of the stack region.  But that's a feature for the future.

Note that I called this new feature a "Python stack" because it originated from the desire to get the stackless mode of uPy working without needing to allocate on the heap, and so the new memory region was exclusively used for the Python stack when calling Python functions.  But it now turned into a generic "scoped allocation" mechanism, so it might be an idea to change the naming to reflect that.

dpgeorge committed:
py/runtime: Move mp_exc_recursion_depth to runtime and rename to raise.

For consistency this helper function is renamed to match the other
exception helpers, and moved to their location in runtime.c.

dpgeorge commented:
@tomlogic this is intended to be a comprehensive solution to #3122 (among other things).  Would you be able to verify that this PR satisfies your requirements?

@stinos from a previous comment of yours, MSVC doesn't support VLA, so perhaps you'd like to comment on what's done here and see if it's useful for MSVC builds.

stinos commented:
Tested this with MSVC and it looks ok, all tests pass. Only problem with it is mp_pystack_free is inline hence when including it from C++ code the compiler gives an error on the implicit conversion of void* to uint8_t*. Which is however allowed in C, so it's not really MicroPython's problem but on the other hand this is the public API; I'll leave it up to you whether to fix this or not, I can workaround on my side easily.

Anyway: I think the usefulness mainly comes from the fact there's now a unified way to allocate on the stack. Also not needing the heap in e.g. mp_call_method_self_n_kw/fun_bc_call is nice (I thought this would improve performanc but couldn't measure any actual improvement though maybe I didn't try hard enough). Are you planning to go further and remove all alloca calls with mp_pystack_alloc and supply config options to select how the latter is actually implemented? Otherwise there is quite some extra code and it's harder in the future to write anything which uses stack allocation..

dpgeorge commented:
@stinos thanks for the testing and feedback.

> Only problem with it is mp_pystack_free is inline hence when including it from C++ code the compiler gives an error on the implicit conversion of void* to uint8_t*.

I've now fixed this.  It's a good idea to try and aid C++ integration where possible.

> Are you planning to go further and remove all alloca calls with mp_pystack_alloc and supply config options to select how the latter is actually implemented?

I tried to cover all alloca() calls in this PR (did I miss any?).  But the code currently looks something like:
```c
#if MICROPY_ENABLE_PYSTACK
void *mem = mp_pystack_alloc(nbytes);
#else
void *mem = alloca(nbytes);
#endif
```
and I guess you are asking if this could be replaced with a (configurable) macro:
```c
void *mem = MP_SCOPED_ALLOC(nbytes);
```
Is that what you mean?

stinos commented:
Yes but I didn't even consider it to be scoped. Actually not being scoped makes it more symmetrical and more like normal C allocations which might be clearer in the end. So instead of  
```
#if MICROPY_ENABLE_PYSTACK
void *mem = mp_pystack_alloc(nbytes);
#else
void *mem = alloca(nbytes);
#endif
//use mem
#if MICROPY_ENABLE_PYSTACK
mp_pystack_free(mem);
#682 
```

I think this is clearer, easier to write and more maintainable:

```
void *mem = mp_pystack_alloc(nbytes);
//use mem
mp_pystack_free(mem);
```

And by means of configuration options, say `MICROPY_PY_PYSTACK_ALLOCA` mp_pystack_alloc would be defined to alloca and mp_pystack_free would be a no-op, and there'd be no difference with the current code.

dpgeorge commented:
> Actually not being scoped makes it more symmetrical and more like normal C allocations which might be clearer in the end.

Well, by scoped I meant that you need to free in the reverse order of allocation (ie LIFO).  Probably "scoped" is not the correct word because the big benefit of the pystack is that you can allocate on the pystack in a function and return the pointer to the caller, and the caller can then free the pystack memory (which is unlike alloca() which must be freed in the same function it was allocated).  That feature opens up quite a few more possibilities for heap-free functionality (eg currently printing a big-int (mpz) requires temporarily allocating on the heap, and that could be changed to use pystack; there are a few cases of this).

Really there are 2 conceptual temporary allocations:
1. temporary allocation within a function/scope: alloca() will work here and this can be replaced by a pystack alloc/free pair.
2. temporary allocation within a function where the allocated pointer is returned to, and freed by, the caller (may not be the direct caller, as long as it's freed in a LIFO manner): alloca() won't work here and currently such cases use the heap (ie m_new/m_free), but these could be changed to use pystack.

Both of these cases could be provided via helper functions with relevant names, eg:
1. mp_pystack_alloc_local(), mp_pystack_free_local() - can be aliased to alloca() and nop (or m_new() and m_free()).
2. mp_pystack_alloc_nonlocal(), mp_pystack_free_nonlocal() - can be aliased to m_new() and m_free().

Note that pystack additionally allows to resize the most recently allocated block (and this feature is already used in this PR).  Probably that would fall under category 2 above as mp_pystack_realloc_local() because there is no alloca() equivalent.

dpgeorge commented:
See also a previous, related idea for a scratch buffer: #596.  The concern there was that it would complicate threading because each thread needs a separate scratch buffer.  In the current PR this is also true because each thread needs its own pystack, and that has been implemented.

stinos commented:
> Well, by scoped I meant that you need to free in the reverse order of allocation (ie LIFO)

Thanks for clearing that out, as you figured I interpreted it as 'memory automatically destroyed at end of scope'.

> Both of these cases could be provided via helper functions with relevant names

Sounds very reasonable. And useful.

dpgeorge committed:
py: Introduce a Python stack for scoped allocation.

This patch introduces the MICROPY_ENABLE_PYSTACK option (disabled by
default) which enables a "Python stack" that allows to allocate and free
memory in a scoped, or Last-In-First-Out (LIFO) way, similar to alloca().

A new memory allocation API is introduced along with this Py-stack.  It
includes both "local" and "nonlocal" LIFO allocation.  Local allocation is
intended to be equivalent to using alloca(), whereby the same function must
free the memory.  Nonlocal allocation is where another function may free
the memory, so long as it's still LIFO.

Follow-up patches will convert all uses of alloca() and VLA to the new
scoped allocation API.  The old behaviour (using alloca()) will still be
available, but when MICROPY_ENABLE_PYSTACK is enabled then alloca() is no
longer required or used.

The benefits of enabling this option are (or will be once subsequent
patches are made to convert alloca()/VLA):
- Toolchains without alloca() can use this feature to obtain correct and
  efficient scoped memory allocation (compared to using the heap instead
  of alloca(), which is slower).
- Even if alloca() is available, enabling the Py-stack gives slightly more
  efficient use of stack space when calling nested Python functions, due to
  the way that compilers implement alloca().
- Enabling the Py-stack with the stackless mode allows for even more
  efficient stack usage, as well as retaining high performance (because the
  heap is no longer used to build and destroy stackless code states).
- With Py-stack and stackless enabled, Python-calling-Python is no longer
  recursive in the C mp_execute_bytecode function.

The micropython.pystack_use() function is included to measure usage of the
Python stack.

dpgeorge committed:
py: Convert all uses of alloca() to use new scoped allocation API.

dpgeorge committed:
extmod/machine_signal: Change VLA to use new scoped allocation API.

dpgeorge committed:
extmod/modure: Convert alloca() to use new scoped allocation API.

dpgeorge committed:
unix: Add support for using the Python stack.

dpgeorge committed:
stm32: Add support for using the Python stack.

dpgeorge committed:
py/runtime: Use the Python stack when building *arg and **kwarg state.

With MICROPY_ENABLE_PYSTACK enabled the following language constructs no
longer allocate on the heap: f(*arg), f(**kwarg).

dpgeorge commented:
Ok, I updated this PR and force-pushed it.  The main change is a refactor to move to an new scoped/LIFO allocation API with the following functions:
```c
// local can be replaced by alloca()
void *mp_local_alloc(size_t n_bytes);
void mp_local_free(void *ptr);

// nonlocal can be replaced by heap allocation/resize/free
void *mp_nonlocal_alloc(size_t n_bytes);
void *mp_nonlocal_realloc(void *ptr, size_t old_n_bytes, size_t new_n_bytes);
void mp_nonlocal_free(void *ptr, size_t n_bytes);
```
Apart from this refactoring pretty much everything else stayed the same.

tomlogic commented:
@dpgeorge Looks like a good idea to me, and would work to replace #3122.

stinos commented:
Nice, thanks.

stinos commented:
@dpgeorge I did some performance tests again with a less artificial usecase (calling methods with *args and **kwargs in the same way used in an actual application, and with a gc heap which is quite populated) and the result is good.. With MICROPY_ENABLE_PYSTACK enabled, the code under test runs 5 times faster. Most of the gains come from not having to loop over all heap blocks, not using alloca has only a small effect. For a single function call the aboslute gain is only in the order of uSec (on a PC) of course, but still, when every of those counts or when used in loops things like this start to matter.

So maybe I'll just enable this per default on the MSVC port. It won't hurt always having this in the CI builds either.

One minor issue with MSVC again: in https://github.com/micropython/micropython/blob/master/extmod/modure.c in the re_split function it complains about passing a `const char **` to mp_local_free which takes `void *`, in mod_ure. This was already a problem earlier, there's a comment explaining why, so I was thinking instead of adding yet again a cast, why not just make the `caps` variable a plain `char **`?

dpgeorge commented:
> I did some performance tests again with a less artificial usecase ... and the result is good

Good news, thanks for sharing.  I also saw some minor improvements in speed and RAM (C stack) usage but didn't quantify it fully.

> So maybe I'll just enable this per default on the MSVC port.

Please go ahead.  I'll also probably enable it on stm32 after I make it so the the pystack can live just under the normal C stack, and they can grow in towards each other.

> One minor issue with MSVC again ...  why not just make the caps variable a plain char **?

That leads to a warning in gcc (in unix coverage build, at least) in passing caps to re1_5_recursiveloopprog.  So let's just cast caps in the call to mp_local_free.  Can you confirm that that fixes it for MSVC?

stinos commented:
Yes `mp_local_free((char**)caps)` works just fine.

dpgeorge commented:
> Yes mp_local_free((char**)caps) works just fine.

Ok, that's done in 1b223a42bf79640d37567c4f427767ee97d45897
